sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -378,8 +406,12 @@ object V2ScanRelationPushDown extends Rule[LogicalPlan] with PredicateHelper wit
       }
       (operation, isPushed && !isPartiallyPushed)
     case s @ Sort(order, _, operation @ ScanOperation(project, filter, sHolder: ScanBuilderHolder))
-        if filter.isEmpty && CollapseProject.canCollapseExpressions(
-          order, project, alwaysInline = true) =>
+        // Without building the Scan, we do not know the resulting column names after aggregate
+        // push-down, and thus can't push down Top-N which needs to know the ordering column names.
+        // TODO: we can support simple cases like GROUP BY columns directly and ORDER BY the same
+        //       columns, which we know the resulting column names: the original table columns.
+        if sHolder.pushedAggregate.isEmpty && filter.isEmpty &&
+          CollapseProject.canCollapseExpressions(order, project, alwaysInline = true) =>
```

Comment:
Hmm, is this regression? I.e. previously we can pushdown, but now we cannot?

Reply:
Previously we can't push down either. After agg pushdown, `Scan` is already built, so limit/offset/topN pushdown won't apply anyway.

Reply:
I see. I was confused by the added comment and thought it was supported before.